node/consensus/src/lib.rs

```diff
@@ -316,32 +340,38 @@ impl<N: Network> Consensus<N> {
         priority_fee: PriorityFee<N>,
         id: N::TransactionID,
         transaction: Transaction<N>,
+        size: TransactionSize,
     ) -> Result<()> {
         // Get the current timestamp.
         let timestamp = now_nanos();
         // Add the Transaction to the transactions_in_queue.
         let inserted = self.transactions_in_queue.lock().insert(id);
         ensure!(inserted, "Transaction '{}' exists in the queue", fmt_id(id));
+        // Add the size to the tracking transactions_queue_size.
+        self.transactions_queue_size.fetch_update(Relaxed, Relaxed, |x| Some(x.saturating_add(size))).map_err(error)?;
         // Add the Transaction to the queue.
-        let found = self.transactions_queue.lock().insert((priority_fee, -timestamp, *id.deref()), transaction);
+        let found = self.transactions_queue.lock().insert((priority_fee, -timestamp, -size, *id.deref()), transaction);
         ensure!(found.is_none(), "Transaction '{}' exists in the memory pool", fmt_id(id));
         Ok(())
     }
 
     /// Pops an unconfirmed transaction from the transaction_queue
-    fn pop_transactions_from_queue(&self, capacity: usize) -> Vec<(TransactionKey<N>, Transaction<N>)> {
+    fn pop_transactions_from_queue(&self, capacity: usize) -> Result<Vec<(TransactionKey<N>, Transaction<N>)>> {
         // Acquire the lock on the queue.
         let mut queue = self.transactions_queue.lock();
         // Determine the number of transactions to send.
         let num_transactions = queue.len().min(capacity);
         // Pop the transactions from the queue.
         let transactions = (0..num_transactions).map(|_| queue.pop_last().unwrap()).collect::<Vec<_>>();
-        // Remove the transactions from the queue index and mark them again freshly as seen.
-        for (_, tx) in transactions.iter() {
+        // Remove the transactions from the queue index, size tracker and mark them again freshly as seen.
+        for ((_, _, negative_size, _), tx) in transactions.iter() {
```

Comment:
why use negative size here instead of subtracting?

Reply:
Because we're storing a negative size, because I want to prioritize ordering smaller transactions as extra DoS protection. It is very ugly, but it makes sense, and is the same ugliness as we already have by storing negative timestamps... Ideas are welcome.

Reply:
if we just want to reverse the sorting order we can use [`cmp::Reverse`](https://doc.rust-lang.org/std/cmp/struct.Reverse.html) instead; it conveys the rationale better IMO

Reply:
TIL, brilliant fee6b4666